subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutor.java

```diff
@@ -139,37 +140,46 @@ private void waitForDeployments(BuildAction action, BuildRequestMetaData request
     private BuildActionRunner.Result executeMultipleBuilds(BuildAction action, BuildRequestMetaData requestContext, BuildSessionContext buildSession,
                                                            BuildCancellationToken cancellationToken, CancellableOperationManager cancellableOperationManager, ContinuousExecutionGate continuousExecutionGate) {
         BuildActionRunner.Result lastResult;
-        while (true) {
-            PendingChangesListener pendingChangesListener = listenerManager.getBroadcaster(PendingChangesListener.class);
-            final FileSystemChangeWaiter waiter = changeWaiterFactory.createChangeWaiter(new SingleFirePendingChangesListener(pendingChangesListener), cancellationToken, continuousExecutionGate);
-            try {
-                lastResult = executeBuildAndAccumulateInputs(action, waiter, buildSession);
+        PendingChangesListener pendingChangesListener = listenerManager.getBroadcaster(PendingChangesListener.class);
+        virtualFileSystem.registerChangeBroadcaster(listenerManager.getBroadcaster(FileChangeListener.class));
+        try {
+            while (true) {
+                FileSystemChangeListener fileSystemChangeListener = new FileSystemChangeListener(
```

Comment:
There is a `FileChangeListener` and a `FileSystemChangeListener`, the latter implementing the former. Can we disambiguate these a bit more?